### PR TITLE
GraphQL - Implement roster category

### PR DESCRIPTION
**Pull request by Premwoik**
This PR addresses [MIM-1605](https://erlangsolutions.atlassian.net/browse/MIM-1605) and implements roster category for both user and admin.

This implementation reuse roster commands from `mod_commands` (used by admin and user Rest API) and extracts those functions to the new `mod_roster_api` module. 

There is also a `service_admin_extra_roster` module that delivers commands for CTL, and I tried to implement the most useful functionality in GraphQL with the use of functions proposed by Rest API. The roster functions used by extra roster commands are undocumented and seem to be older. Also, the extra roster API is a little bit different from the Rest API, so I left it unchanged.

### CODE REVIEW

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/113918) / small_tests / 6ab1611ed9988e541bb14e675104f3ea4c93b787
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3586/113918/small_tests.24.0.5) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3586/113918/small_tests.24.0.5/small/ct_run.mongooseim@localhost.2022-03-14_08.08.56/lib.mongooseim.logs/run.2022-03-14_08.08.56/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/113941) / small_tests / a4475ee2675367af1a302dab5d9876d13033ef57
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3586/113941/small_tests.24.0.5) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3586/113941/small_tests.24.0.5/small/ct_run.mongooseim@localhost.2022-03-14_08.48.53/lib.mongooseim.logs/run.2022-03-14_08.48.53/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/114016) / small_tests / bb0ab5756b0cf26dda4f336275cfd1550675bf4e
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3586/114016/small_tests.24.0.5) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3586/114016/small_tests.24.0.5/small/ct_run.mongooseim@localhost.2022-03-14_11.21.04/lib.mongooseim.logs/run.2022-03-14_11.21.04/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/114063) / small_tests / 7a6aa7d91b44c03c2a8d25dec8c0e2e8b01bfd6e
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3586/114063/small_tests.24.0.5) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3586/114063/small_tests.24.0.5/small/ct_run.mongooseim@localhost.2022-03-14_12.51.53/lib.mongooseim.logs/run.2022-03-14_12.51.53/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/114091) / small_tests / ab8ad9331c547f76dfb0a442f5d171e591ada7d7
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3586/114091/small_tests.24.0.5) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3586/114091/small_tests.24.0.5/small/ct_run.mongooseim@localhost.2022-03-14_19.55.42/lib.mongooseim.logs/run.2022-03-14_19.55.42/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/114110) / small_tests / 2c6ec682657ba8867accefcbc070852c2aabcf0c
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3586/114110/small_tests.24.0.5) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3586/114110/small_tests.24.0.5/small/ct_run.mongooseim@localhost.2022-03-14_20.58.36/lib.mongooseim.logs/run.2022-03-14_20.58.37/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/114172) / small_tests / f4e7d4e741e5ba4105c2b610e93bc33f03732ce7
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3586/114172/small_tests.24.0.5) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3586/114172/small_tests.24.0.5/small/ct_run.mongooseim@localhost.2022-03-14_21.57.34/lib.mongooseim.logs/run.2022-03-14_21.57.34/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/114196) / small_tests / 120d43a0e7c1e1b5e63ebc453e6e85cdc892c547
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3586/114196/small_tests.24.0.5) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3586/114196/small_tests.24.0.5/small/ct_run.mongooseim@localhost.2022-03-14_22.20.06/lib.mongooseim.logs/run.2022-03-14_22.20.07/suite.log.html)

File: src/mod_roster_api.erl [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+           fun() -> subscription(UserA, UserB, subscribe) end,
+           fun() -> subscription(UserB, UserA, subscribe) end,
+           fun() -> subscription(UserA, UserB, subscribed) end,
+           fun() -> subscription(UserB, UserA, subscribed) end],
+    case run_seq(Seq, ok) of
+        ok ->
+            {ok, io_lib:format("Subscription between users ~p and ~p created successfully",
+                               [jid:to_binary(UserA), jid:to_binary(UserB)])};
+        Error ->
+           Error
```
Comment (chrzaszcz) [inline src/mod_roster_api.erl:143]:
```suggestion
            Error
```

File: priv/graphql/schemas/admin/roster.gql [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+  subscription(user: JID!, contact: JID!, action: SubAction): String
+  "Delete user's contact"
+  deleteContact(user: JID!, contact: JID!): String
+  "Delete user's contacts"
+  deleteContacts(user: JID!, contacts: [JID!]!): [String]!
+  "Manage mutual subscription between given users"
+  setMutualSubscription(userA: JID!, userB: JID!, action: MutualSubAction!): String
+  "Set both subscritpion of user to users from contact list"
+  subscribeToAll(user: ContactInput!, contacts: [ContactInput!]!): [String]!
+  "Subscibe all users to all users from the given contacts list"
```
Comment (chrzaszcz) [inline priv/graphql/schemas/admin/roster.gql:19]:
```suggestion
  "Set mutual subscriptions between all of the given contacts"
```

**Review by chrzaszcz — COMMENTED** [review]
Looks good, a huge amount of good code 👍

File: src/graphql/admin/mongoose_graphql_admin_mutation.erl [hunk 3]
```diff
@@ -16,4 +16,6 @@ execute(_Ctx, _Obj, <<"muc_light">>, _Args) ->
 execute(_Ctx, _Obj, <<"session">>, _Args) ->
     {ok, session};
 execute(_Ctx, _Obj, <<"stanza">>, _Args) ->
-    {ok, #{}}.
+    {ok, #{}};
```
[the remark this answers is not in the document]
Comment (Premwoik) [reply src/graphql/admin/mongoose_graphql_admin_mutation.erl:19]: I will change it to `stanza` to follow the others. This value is passed under the `Obj` attribute to the `execute` function that resolves the `StanzaAdminMutation` type. It's useful when you want to pass on some information from the parent. Currently, we don't use `Obj` arg. Probably Michael put here a map and I put a parent's name in other categories.

File: big_tests/tests/graphql_roster_SUITE.erl [hunk 4]
```diff
@@ -0,0 +1,657 @@
...
+
+admin_subscribe_to_all(Config) ->
+    escalus:fresh_story_with_config(Config, [{alice, 1}, {bob, 1}, {kate, 1}],
+                                    fun admin_subscribe_to_all_story/4).
+
+admin_subscribe_to_all_story(Config, Alice, Bob, Kate) ->
+    Res = execute_auth(admin_subscribe_to_all_body(Alice, [Bob, Kate]), Config),
+    check_if_created_succ(?SUBSCRIBE_TO_ALL_PATH, Res),
+
+    check_contacts([Bob, Kate], Alice, Config),
```
[the remark this answers is not in the document]
Comment (Premwoik) [reply big_tests/tests/graphql_roster_SUITE.erl:262]: Ok, I will use `check_contacts` in `...contacts...` tests. I probably introduce the `check_contacts` function after I created that test cases.

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/115133) / small_tests / 1c8e5869f926cee28f6d32916e21e3a566af9c66
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3586/115133/small_tests.24.0.5) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3586/115133/small_tests.24.0.5/small/ct_run.mongooseim@localhost.2022-03-17_17.16.34/lib.mongooseim.logs/run.2022-03-17_17.16.34/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/115252) / small_tests / 364b3211a2c0de2cee47eba3c90954cf28f6aecd
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3586/115252/small_tests.24.0.5) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3586/115252/small_tests.24.0.5/small/ct_run.mongooseim@localhost.2022-03-17_22.50.24/lib.mongooseim.logs/run.2022-03-17_22.50.24/suite.log.html)

**Review by chrzaszcz — APPROVED** [review]
All looking good!

**Pull request merged**